src/hotspot/cpu/x86/assembler_x86.cpp

```diff
@@ -12799,13 +13358,66 @@ int Assembler::prefix_and_encode(int dst_enc, bool dst_is_byte, int src_enc, boo
   return dst_enc << 3 | src_enc;
 }
 
-int8_t Assembler::get_prefixq(Address adr) {
+void Assembler::prefix16(int prefix) {
+  emit_int8((prefix & 0xff00) >> 8);
+  emit_int8(prefix & 0xff);
+}
+
+int Assembler::prefix_and_encode_rex2(int dst_enc, bool dst_is_byte, int src_enc, bool src_is_byte) {
+  if (dst_enc >= 24) {
+    dst_enc -= 24;
+    if (src_enc >= 24) {
+      src_enc -= 24;
+      prefix16(REX2_R4B4RB);
+    }
+    else if (src_enc >= 16) {
+      src_enc -= 16;
+      prefix16(REX2_R4B4R);
+    }
+  }
+  else if (dst_enc >= 16) {
+    dst_enc -= 16;
+    if (src_enc >= 24) {
+      src_enc -= 24;
+      prefix16(REX2_R4B4B);
+    }
+    else if (src_enc >= 16) {
+      src_enc -= 16;
+      prefix16(REX2_R4B4);
+    }
+  }
+  return dst_enc << 3 | src_enc;
+}
+
+bool Assembler::prefix_is_rex2(int prefix) {
+  return (prefix & 0xFF00) == REX2;
+}
+
+int Assembler::get_prefixq(Address adr, bool isPage1) {
+  if (adr.base_needs_rex2() || adr.index_needs_rex2()) {
+    return get_prefixq_rex2(adr, isPage1);  
+  }
   int8_t prfx = get_prefixq(adr, rax);
-  assert(REX_W <= prfx && prfx <= REX_WXB, "must be");
-  return prfx;
+  assert(REX_W <= prfx && prfx <= REX2_X4B4W, "must be");
+  return (int16_t)prfx;
```

Comment:
This could be changed to:
return isPage1 ? (((int16_t)prfx) <<  8) | 0x0F, (int16_t)prfx;

Reply:
Very nice!  Done.